src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs

```diff
@@ -409,13 +440,99 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Cancellat
+                    // on the current snapshot. So, we deem this as an important analyzer that should not be de-prioritized here.
+                    // Note that we only perform this analysis if the prior document, whose existingData is cached, had same number
+                    // of source lines as the current document snapshot. Otherwise, the start/end lines comparison across
+                    // snapshots is not meaningful.
+                    if (existingData.LineCount == _text.Lines.Count &&
+                        !existingData.Items.IsEmpty)
+                    {
+                        _text.GetLinesAndOffsets(span.Value, out var startLineNumber, out var _, out var endLineNumber, out var _);
+
+                        foreach (var diagnostic in existingData.Items)
+                        {
+                            if (diagnostic.DataLocation.MappedFileSpan.StartLinePosition.Line <= endLineNumber &&
+                                diagnostic.DataLocation.MappedFileSpan.EndLinePosition.Line >= startLineNumber)
+                            {
+                                return false;
+                            }
+                        }
+                    }
+
+                    // 'LightbulbSkipExecutingDeprioritizedAnalyzers' option determines if we want to execute this analyzer
+                    // in low priority bucket or skip it completely. If the option is not set, track the de-prioritized
+                    // analyzer to be executed in low priority bucket.
+                    if (!_owner.GlobalOptions.GetOption(DiagnosticOptionsStorage.LightbulbSkipExecutingDeprioritizedAnalyzers))
+                        _priorityProvider.AddDeprioritizedAnalyzerWithLowPriority(analyzer);
+
+                    return true;
+                }
+
+                // Returns true if this is an analyzer that is a candidate to be de-prioritized to
+                // 'CodeActionRequestPriority.Low' priority for improvement in analyzer
+                // execution performance for priority buckets above 'Low' priority.
+                // Based on performance measurements, currently only analyzers which register SymbolStart/End actions
+                // or SemanticModel actions are considered candidates to be de-prioritized. However, these semantics
+                // could be changed in future based on performance measurements.
+                async Task<bool> IsCandidateForDeprioritizationBasedOnRegisteredActionsAsync(DiagnosticAnalyzer analyzer)
+                {
+                    // We deprioritize SymbolStart/End and SemanticModel analyzers from 'Normal' to 'Low' priority bucket,
+                    // as these are computationally more expensive.
+                    // Note that we never de-prioritize compiler analyzer, even though it registers a SemanticModel action.
+                    if (_compilationWithAnalyzers == null ||
```

Comment:
what is the `_compilationWithAnalyzers  == null` check for?  when does that happen?

Reply:
There are couple of places where that can happen, and we have to keep hold of this as a nullable field/parameter throughout the IDE code: https://github.com/dotnet/roslyn/blob/c24e8dafec132217b970904758eca56f35318aba/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs#L18-L58